In [3]:
import numpy as np
import pandas as pd

### Question 1

What's the version of NumPy that you installed? 

You can get the version information using the `__version__` field:

```python
np.__version__
```

### Getting the data 

For this homework, we'll use the Car price dataset. Download it from 
[here](https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv).

You can do it with wget:

```bash
wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
```

Or just open it with your browser and click "Save as...".

Now read it with Pandas.

In [4]:
np.__version__

'1.22.4'

In [3]:
!wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv

--2023-03-29 09:26:17--  https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.110.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1475504 (1,4M) [text/plain]
Saving to: ‘data.csv’

data.csv            100%[===================>]   1,41M  7,79MB/s    in 0,2s    

2023-03-29 09:26:17 (7,79 MB/s) - ‘data.csv’ saved [1475504/1475504]



In [5]:
df = pd.read_csv("data.csv")

In [6]:
df

Make       Model  Year                Engine Fuel Type  Engine HP  \
0          BMW  1 Series M  2011     premium unleaded (required)      335.0   
1          BMW    1 Series  2011     premium unleaded (required)      300.0   
2          BMW    1 Series  2011     premium unleaded (required)      300.0   
3          BMW    1 Series  2011     premium unleaded (required)      230.0   
4          BMW    1 Series  2011     premium unleaded (required)      230.0   
...        ...         ...   ...                             ...        ...   
11909    Acura         ZDX  2012     premium unleaded (required)      300.0   
11910    Acura         ZDX  2012     premium unleaded (required)      300.0   
11911    Acura         ZDX  2012     premium unleaded (required)      300.0   
11912    Acura         ZDX  2013  premium unleaded (recommended)      300.0   
11913  Lincoln      Zephyr  2006                regular unleaded      221.0   

       Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
0                   6.0            MANUAL   rear wheel drive              2.0   
1                   6.0            MANUAL   rear wheel drive              2.0   
2                   6.0            MANUAL   rear wheel drive              2.0   
3                   6.0            MANUAL   rear wheel drive              2.0   
4                   6.0            MANUAL   rear wheel drive              2.0   
...                 ...               ...                ...              ...   
11909               6.0         AUTOMATIC    all wheel drive              4.0   
11910               6.0         AUTOMATIC    all wheel drive              4.0   
11911               6.0         AUTOMATIC    all wheel drive              4.0   
11912               6.0         AUTOMATIC    all wheel drive              4.0   
11913               6.0         AUTOMATIC  front wheel drive              4.0   

                             Market Category Vehicle Size  Vehicle Style  \
0      Factory Tuner,Luxury,High-Performance      Compact          Coupe   
1                         Luxury,Performance      Compact    Convertible   
2                    Luxury,High-Performance      Compact          Coupe   
3                         Luxury,Performance      Compact          Coupe   
4                                     Luxury      Compact    Convertible   
...                                      ...          ...            ...   
11909             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11910             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11911             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11912             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11913                                 Luxury      Midsize          Sedan   

       highway MPG  city mpg  Popularity   MSRP  
0               26        19        3916  46135  
1               28        19        3916  40650  
2               28        20        3916  36350  
3               28        18        3916  29450  
4               28        18        3916  34500  
...            ...       ...         ...    ...  
11909           23        16         204  46120  
11910           23        16         204  56670  
11911           23        16         204  50620  
11912           23        16         204  50920  
11913           26        17          61  28995  

[11914 rows x 16 columns]

In [7]:
df.describe()

Year    Engine HP  Engine Cylinders  Number of Doors  \
count  11914.000000  11845.00000      11884.000000     11908.000000   
mean    2010.384338    249.38607          5.628829         3.436093   
std        7.579740    109.19187          1.780559         0.881315   
min     1990.000000     55.00000          0.000000         2.000000   
25%     2007.000000    170.00000          4.000000         2.000000   
50%     2015.000000    227.00000          6.000000         4.000000   
75%     2016.000000    300.00000          6.000000         4.000000   
max     2017.000000   1001.00000         16.000000         4.000000   

        highway MPG      city mpg    Popularity          MSRP  
count  11914.000000  11914.000000  11914.000000  1.191400e+04  
mean      26.637485     19.733255   1554.911197  4.059474e+04  
std        8.863001      8.987798   1441.855347  6.010910e+04  
min       12.000000      7.000000      2.000000  2.000000e+03  
25%       22.000000     16.000000    549.000000  2.100000e+04  
50%       26.000000     18.000000   1385.000000  2.999500e+04  
75%       30.000000     22.000000   2009.000000  4.223125e+04  
max      354.000000    137.000000   5657.000000  2.065902e+06

### Question 2

How many records are in the dataset?

Here you need to specify the number of rows.

- 16
- 6572
- 11914
- 18990

## Ans = 11914

In [13]:
df['Make'].value_counts()

Chevrolet        1123
Ford              881
Volkswagen        809
Toyota            746
Dodge             626
Nissan            558
GMC               515
Honda             449
Mazda             423
Cadillac          397
Mercedes-Benz     353
Suzuki            351
BMW               334
Infiniti          330
Audi              328
Hyundai           303
Volvo             281
Subaru            256
Acura             252
Kia               231
Mitsubishi        213
Lexus             202
Buick             196
Chrysler          187
Pontiac           186
Lincoln           164
Oldsmobile        150
Land Rover        143
Porsche           136
Saab              111
Aston Martin       93
Plymouth           82
Bentley            74
Ferrari            69
FIAT               62
Scion              60
Maserati           58
Lamborghini        52
Rolls-Royce        31
Lotus              29
Tesla              18
HUMMER             17
Maybach            16
Alfa Romeo          5
McLaren             5
Spyker    

### Question 3

Who are the most frequent car manufacturers (top-3) according to the dataset?

- Chevrolet, Volkswagen, Toyota
- Chevrolet, Ford, Toyota
- Ford, Volkswagen, Toyota
- Chevrolet, Ford, Volkswagen

## Ans = Chevrolet, Ford, Volkswagen

In [16]:
df_audi = df[df['Make'] == 'Audi']

In [17]:
df_audi

Make Model  Year             Engine Fuel Type  Engine HP  \
17     Audi   100  1992             regular unleaded      172.0   
18     Audi   100  1992             regular unleaded      172.0   
19     Audi   100  1992             regular unleaded      172.0   
20     Audi   100  1992             regular unleaded      172.0   
21     Audi   100  1992             regular unleaded      172.0   
...     ...   ...   ...                          ...        ...   
10901  Audi   TTS  2016  premium unleaded (required)      292.0   
10902  Audi   TTS  2017  premium unleaded (required)      292.0   
11178  Audi    V8  1992             regular unleaded      276.0   
11179  Audi    V8  1993             regular unleaded      276.0   
11180  Audi    V8  1994             regular unleaded      276.0   

       Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
17                  6.0            MANUAL  front wheel drive              4.0   
18                  6.0            MANUAL  front wheel drive              4.0   
19                  6.0         AUTOMATIC    all wheel drive              4.0   
20                  6.0            MANUAL  front wheel drive              4.0   
21                  6.0            MANUAL    all wheel drive              4.0   
...                 ...               ...                ...              ...   
10901               4.0  AUTOMATED_MANUAL    all wheel drive              2.0   
10902               4.0  AUTOMATED_MANUAL    all wheel drive              2.0   
11178               8.0         AUTOMATIC    all wheel drive              4.0   
11179               8.0         AUTOMATIC    all wheel drive              4.0   
11180               8.0         AUTOMATIC    all wheel drive              4.0   

                             Market Category Vehicle Size Vehicle Style  \
17                                    Luxury      Midsize         Sedan   
18                                    Luxury      Midsize         Sedan   
19                                    Luxury      Midsize         Wagon   
20                                    Luxury      Midsize         Sedan   
21                                    Luxury      Midsize         Sedan   
...                                      ...          ...           ...   
10901       Factory Tuner,Luxury,Performance      Compact         Coupe   
10902  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
11178                     Luxury,Performance      Midsize         Sedan   
11179                     Luxury,Performance      Midsize         Sedan   
11180                     Luxury,Performance      Midsize         Sedan   

       highway MPG  city mpg  Popularity   MSRP  
17              24        17        3105   2000  
18              24        17        3105   2000  
19              20        16        3105   2000  
20              24        17        3105   2000  
21              21        16        3105   2000  
...            ...       ...         ...    ...  
10901           27        23        3105  51900  
10902           27        23        3105  52500  
11178           18        13        3105   2000  
11179           18        13        3105   2149  
11180           18        13        3105   2239  

[328 rows x 16 columns]

In [18]:
df_audi['Model'].nunique()

34

### Question 4

What's the number of unique Audi car models in the dataset?

- 3
- 16
- 26
- 34

## Ans = 34

In [19]:
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

### Question 5

How many columns in the dataset have missing values?

- 5
- 6
- 7
- 8

## Ans = 5

In [21]:
df['Engine Cylinders'].median()

6.0

In [23]:
df['Engine Cylinders'].value_counts()

4.0     4752
6.0     4489
8.0     2031
12.0     230
5.0      225
10.0      68
0.0       56
3.0       30
16.0       3
Name: Engine Cylinders, dtype: int64

In [24]:
df['Engine Cylinders'].isnull().sum()

30

In [27]:
df['Engine Cylinders'].fillna(4752).isnull().sum()

0

In [26]:
df['Engine Cylinders'].isnull().sum()

30

In [30]:
df[['Engine Cylinders']] = df[['Engine Cylinders']].fillna(4752)

In [31]:
df

Make       Model  Year                Engine Fuel Type  Engine HP  \
0          BMW  1 Series M  2011     premium unleaded (required)      335.0   
1          BMW    1 Series  2011     premium unleaded (required)      300.0   
2          BMW    1 Series  2011     premium unleaded (required)      300.0   
3          BMW    1 Series  2011     premium unleaded (required)      230.0   
4          BMW    1 Series  2011     premium unleaded (required)      230.0   
...        ...         ...   ...                             ...        ...   
11909    Acura         ZDX  2012     premium unleaded (required)      300.0   
11910    Acura         ZDX  2012     premium unleaded (required)      300.0   
11911    Acura         ZDX  2012     premium unleaded (required)      300.0   
11912    Acura         ZDX  2013  premium unleaded (recommended)      300.0   
11913  Lincoln      Zephyr  2006                regular unleaded      221.0   

       Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
0                   6.0            MANUAL   rear wheel drive              2.0   
1                   6.0            MANUAL   rear wheel drive              2.0   
2                   6.0            MANUAL   rear wheel drive              2.0   
3                   6.0            MANUAL   rear wheel drive              2.0   
4                   6.0            MANUAL   rear wheel drive              2.0   
...                 ...               ...                ...              ...   
11909               6.0         AUTOMATIC    all wheel drive              4.0   
11910               6.0         AUTOMATIC    all wheel drive              4.0   
11911               6.0         AUTOMATIC    all wheel drive              4.0   
11912               6.0         AUTOMATIC    all wheel drive              4.0   
11913               6.0         AUTOMATIC  front wheel drive              4.0   

                             Market Category Vehicle Size  Vehicle Style  \
0      Factory Tuner,Luxury,High-Performance      Compact          Coupe   
1                         Luxury,Performance      Compact    Convertible   
2                    Luxury,High-Performance      Compact          Coupe   
3                         Luxury,Performance      Compact          Coupe   
4                                     Luxury      Compact    Convertible   
...                                      ...          ...            ...   
11909             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11910             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11911             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11912             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11913                                 Luxury      Midsize          Sedan   

       highway MPG  city mpg  Popularity   MSRP  
0               26        19        3916  46135  
1               28        19        3916  40650  
2               28        20        3916  36350  
3               28        18        3916  29450  
4               28        18        3916  34500  
...            ...       ...         ...    ...  
11909           23        16         204  46120  
11910           23        16         204  56670  
11911           23        16         204  50620  
11912           23        16         204  50920  
11913           26        17          61  28995  

[11914 rows x 16 columns]

In [32]:
df['Engine Cylinders'].isnull().sum()

0

In [33]:
df['Engine Cylinders'].median()

6.0

In [34]:
df['Engine Cylinders'].value_counts()

4.0       4752
6.0       4489
8.0       2031
12.0       230
5.0        225
10.0        68
0.0         56
4752.0      30
3.0         30
16.0         3
Name: Engine Cylinders, dtype: int64

### Question 6

1. Find the median value of "Engine Cylinders" column in the dataset.
2. Next, calculate the most frequent value of the same "Engine Cylinders".
3. Use the `fillna` method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
4. Now, calculate the median value of "Engine Cylinders" once again.

Has it changed?

> Hint: refer to existing `mode` and `median` functions to complete the task.

- Yes
- No

## Ans = No

In [35]:
list(df)

['Make',
 'Model',
 'Year',
 'Engine Fuel Type',
 'Engine HP',
 'Engine Cylinders',
 'Transmission Type',
 'Driven_Wheels',
 'Number of Doors',
 'Market Category',
 'Vehicle Size',
 'Vehicle Style',
 'highway MPG',
 'city mpg',
 'Popularity',
 'MSRP']

In [42]:
df_temp = df[df['Make'] == 'Lotus'][["Engine HP", "Engine Cylinders"]]

In [43]:
len(df_temp)

29

In [44]:
df_temp

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3914      189.0               4.0
3915      189.0               4.0
3916      218.0               4.0
3917      189.0               4.0
3918      217.0               4.0
3919      217.0               4.0
4216      350.0               8.0
4217      350.0               8.0
4218      350.0               8.0
4257      400.0               6.0
4258      400.0               6.0
4259      276.0               6.0
4260      276.0               6.0
4261      276.0               6.0
4262      345.0               6.0
4263      345.0               6.0
4264      276.0               6.0
4265      345.0               6.0
4266      276.0               6.0
4267      345.0               6.0
4268      276.0               6.0
4292      257.0               4.0
4293      240.0               4.0
4294      240.0               4.0
4295      240.0               4.0
4296      257.0               4.0
4297      257.0               4.0

In [49]:
df_temp.drop_duplicates().reset_index(drop=True)

Engine HP  Engine Cylinders
0      189.0               4.0
1      218.0               4.0
2      217.0               4.0
3      350.0               8.0
4      400.0               6.0
5      276.0               6.0
6      345.0               6.0
7      257.0               4.0
8      240.0               4.0

In [65]:
X = df_temp.drop_duplicates().reset_index(drop=True).values

In [66]:
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [67]:
XT = np.transpose(X)
XT

array([[189., 218., 217., 350., 400., 276., 345., 257., 240.],
       [  4.,   4.,   4.,   8.,   6.,   6.,   6.,   4.,   4.]])

In [68]:
XTX = XT.dot(X)

In [69]:
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [70]:
XTX.shape

(2, 2)

In [71]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [72]:
print(XTX_inv.shape, XT.shape)

(2, 2) (2, 9)


In [74]:
y = [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]

temp = XTX_inv.dot(XT)

w = temp.dot(y)

In [75]:
w

array([  4.59494481, -63.56432501])

### Question 7

1. Select all the "Lotus" cars from the dataset.
2. Select only columns "Engine HP", "Engine Cylinders".
3. Now drop all duplicated rows using `drop_duplicates` method (you should get a dataframe with 9 rows).
4. Get the underlying NumPy array. Let's call it `X`.
5. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
6. Compute the inverse of `XTX`.
7. Create an array `y` with values `[1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]`.
8. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
9. What's the value of the first element of `w`?

> **Note**: You just implemented linear regression. We'll talk about it in the next lesson.

- -0.0723
- 4.5949
- 31.6537
- 63.5643

# Ans = 4.5949